pkg/azuredx/azureauth/auth.go

```diff
@@ -0,0 +1,133 @@
+// Package azureauth provides authorization utilities for the Azure cloud.
+package azureauth
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/azure-data-explorer-datasource/pkg/azuredx/models"
+)
+
+var onBehalfOfLatencySeconds = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+	Namespace: "grafana",
+	Subsystem: "plugin_adx",
+	Name:      "obo_latency_seconds",
+	Help:      "On-behalf-of token exchange duration.",
+	Buckets:   []float64{0.01, 0.05, 0.10, 0.20, 0.40, 1.00},
+}, []string{"http_status"})
+
+func init() {
+	prometheus.MustRegister(onBehalfOfLatencySeconds)
+}
+
+// ServiceCredentials provides authorization for cloud service usage.
+type ServiceCredentials struct {
+	models.DatasourceSettings
+	// PostForm is an http.Client method.
+	PostForm func(url string, data url.Values) (*http.Response, error)
+}
+
+// QueryDataAuthorization returns an HTTP Authorization-header value which
+// represents the respective request.
+// ⚠️ A zero string must cause appliance of the service account instead.
+//
+// TODO(pascaldekloe): Merge the service principal appliance in here. The entire
+//	authorization decission must be centralised for security reasons alone.
+//	A fully progmatic approach also bypasses the SDK's incorrect use of
+//	http.RoundTripper.
+func (c *ServiceCredentials) QueryDataAuthorization(req *backend.QueryDataRequest) (string, error) {
```

Comment:
`QueryData...` implies that this function does the query, I would rename it to `GetOnBehalfOfBearerToken`. Also, I would set the settings and http client as parameters so you don't need the struct (see my comment below in datasource.go:66). Something like: `GetOnBehalfOfBearerToken(settings, req, http.Client)`.

Reply:
QueryDataAuthorization gets you the Authorization of QueryData, just like IceCreamCount gets you the Count of IceCream (and not the IceCream itself).

Again, function names should describe what they do—now how. The actual methods like on-behalf-of and/or bearer-tokens are likely to fluctuate over time.

You *never* need a struct, as everything can be done by passing arguments to functions. The struct here encapsulates a responsibility with easy to understand boundaries. The method call clearly expresses that it provides the authorization for a request.